File: trunk/src/m/contrib/defleurian/netCDF/export_netCDF.m
function export_netCDF(md,filename)	
	
%Now going on Real treatment
	if exist(filename),
		disp(sprintf('File %s allready exist', filename));
		prompt = 'Give a new name or "delete" to replace: ';
		newname = input(prompt,'s');
		if strcmp(newname,'delete')
			delete(filename)
		else
			disp(sprintf('New file name is %s ', newname));
			filename=newname
	  end
  end
	%open file and write description
	mode = netcdf.getConstant('NC_NETCDF4');
	mode = bitor(mode,netcdf.getConstant('NC_NOCLOBBER'));%NOCLOBBER to avoid overwrite
	ncid = netcdf.create(filename,mode);
	netcdf.putAtt(ncid,netcdf.getConstant('NC_GLOBAL'),'Title',['Results for run ' md.miscellaneous.name]);
	netcdf.putAtt(ncid,netcdf.getConstant('NC_GLOBAL'),'Date',['Created ' datestr(now)]);
	
	%gather geometry and timestepping as dimensions
	resfields=fieldnames(md.results);
	Duration=size(eval(['md.results. ' resfields{1} ]),2);
	if Duration>0,
		StepNum=Duration;
	else
		StepNum=1;
  end							

   dimlist=[40,2,md.mesh.numberofelements,md.mesh.numberofvertices,size(md.mesh.elements,2)];
 
	%define netcdf dimensions
	DimSize(1).index=netcdf.defDim(ncid,'Dimension1',StepNum);
	[DimSize(1).name,DimSize(1).value]=netcdf.inqDim(ncid,DimSize(1).index);
	DimValue(1)=DimSize(1).value;
	for i=1:5
		if sum(dimlist(i) == DimValue) == 0
			DimSize(i+1).index=netcdf.defDim(ncid,['Dimension' num2str(i+1)],dimlist(i));
			[DimSize(i+1).name,DimSize(i+1).value]=netcdf.inqDim(ncid,DimSize(i+1).index);
			DimValue(i+1)=DimSize(i+1).value;
		end
	end

	typelist=[{'numeric'} {'logical'} {'string'} {'char'} {'cell'}];
 
	%get all model classes and create respective groups
	groups=fieldnames(md);
	for i=1:length(groups),
		disp(sprintf('group name in tree %s ',groups{i}));
		groupID=netcdf.defGrp(ncid,groups{i});
		%In each group gather the fields of the class
		groupfields=fields(md.(groups{i}));
		for j=1:length(groupfields),
			Var=md.(groups{i}).(groupfields{j});
			if isa(Var,'cell')
				Stdlist=false;
				if length(Var) == 0
					Stdlist=true;
				else
					for k=1:length(typelist)
						if isa(Var{1},typelist{k})
							Stdlist=true;
						end
					end
				end

				netcdf.putAtt(groupID,netcdf.getConstant('NC_GLOBAL'),'classtype',class(md.(groups{i})));
				if(Stdlist)
					disp(sprintf('=====Field name in tree %s ',groupfields{j}));
					[DimSize,DimValue]=DefCreateVar(ncid,Var,groupID,groupfields{j},DimSize,DimValue);
				else
					listsize=length(Var);
					subgroupID=netcdf.defGrp(groupID,groupfields{j});
					netcdf.putAtt(subgroupID,netcdf.getConstant('NC_GLOBAL'),'classtype',class(Var));
					for l=1:listsize
						if isprop(Var{l},'name')
							lname=Var{l}.name;
						elseif isprop(Var{l},'step')
							lname=Var{l}.step
						else 
							lname=[class(Var{l}) int2str(l)];
						end
						listgroupID=netcdf.defGrp(subgroupID,lname);
						netcdf.putAtt(listgroupID,netcdf.getConstant('NC_GLOBAL'),'classtype',class(Var{l}));
						subfields=fields(Var{l});
						for m=1:length(subfields)
							if ~strcmp(subfields{m},'outlog')
								[DimSize,DimValue]=DefCreateVar(ncid,Var{l}.(subfields{m}),listgroupID,subfields{m},DimSize,DimValue);
							end
						end
					end
				end
			elseif isa(Var,'struct') && ~strcmp(groupfields{j},'bamg')
				classtype=class(md.(groups{i}));
				if strcmp(classtype,'struct')
					classtype=groups{i};
				end
				netcdf.putAtt(groupID,netcdf.getConstant('NC_GLOBAL'),'classtype',classtype);
				if length(Var)>1
					listsize=length(Var);
					subgroupID=netcdf.defGrp(groupID,groupfields{j});
					classtype=class(Var);
					if strcmp(classtype,'struct')
						classtype=groups{i};
					end
					netcdf.putAtt(subgroupID,netcdf.getConstant('NC_GLOBAL'),'classtype',classtype);
					for l=1:listsize
						if isfield(Var(l),'step')
							lname=[int2str(Var(l).step)];
						else
							lname=[class(Var(l)) int2str(l)];
						end
						classtype=class(Var(l));
						if strcmp(classtype,'struct')
							classtype=groups{i};
						end
						listgroupID=netcdf.defGrp(subgroupID,lname);
						netcdf.putAtt(listgroupID,netcdf.getConstant('NC_GLOBAL'),'classtype',classtype);
						subfields=fields(Var(l));
						for m=1:length(subfields)
							if ~strcmp(subfields{m},'outlog')
								[DimSize,DimValue]=DefCreateVar(ncid,Var(l).(subfields{m}),listgroupID,subfields{m},DimSize,DimValue);
							end
						end
					end
				else
					subgroupID=netcdf.defGrp(groupID,groupfields{j});
					classtype=class(Var);
					if strcmp(classtype,'struct')
						classtype=groups{i};
					end
					netcdf.putAtt(subgroupID,netcdf.getConstant('NC_GLOBAL'),'classtype',classtype);
					subfields=fields(Var);
					for m=1:length(subfields)
						if ~strcmp(subfields{m},'outlog')
							[DimSize,DimValue]=DefCreateVar(ncid,Var.(subfields{m}),subgroupID,subfields{m},DimSize,DimValue);
						end
					end
				end
			else
				netcdf.putAtt(groupID,netcdf.getConstant('NC_GLOBAL'),'classtype',class(md.(groups{i})));
				[DimSize,DimValue]=DefCreateVar(ncid,Var,groupID,groupfields{j},DimSize,DimValue);
			end
		end
 end
 netcdf.close(ncid);
end

function [DimSize,DimValue]=DefCreateVar(ncid,Var,groupID,field,DimSize,DimValue,last,md,midfield)
	varclass=class(Var);
	varsize=size(Var);
	varlength=length(Var);
	if isa(Var,'logical'),
		if Var,
			LogicString='True';
		else,
			LogicString='False';
  	end
		netcdf.putAtt(groupID,netcdf.getConstant('NC_GLOBAL'),field,LogicString);
	elseif isa(Var,'char'),
		netcdf.putAtt(groupID,netcdf.getConstant('NC_GLOBAL'),field,Var);
	elseif isa(Var,'double'), %dealing with arrays
		[dims,DimSize,DimValue]=GetDims(ncid,Var,DimSize,DimValue);
 		varid = netcdf.defVar(groupID,field,'NC_DOUBLE',dims);
		if length(Var)==0,
			netcdf.putVar(groupID,varid,NaN);
		else
			netcdf.putVar(groupID,varid,Var);
		end
	elseif isa(Var,'cell'),
		[dims,DimSize,DimValue]=GetDims(ncid,Var,DimSize,DimValue);
		%dirty hack to be able to pass strings
		varid = netcdf.defVar(groupID,field,'NC_CHAR',dims);
		if length(Var)==0,
			netcdf.putVar(groupID,varid,0,9,'emptycell')
		else
			for i=1:length(Var),
				if length(Var)>1,
					endpoint=[1,min(length(Var{i}),40)];
					startpoint=[1 0];
				else
					endpoint=min(length(Var{i}),40);
					startpoint=0;
				end
				if length(Var{i})>40,
					netcdf.putVar(groupID,varid,startpoint,extent,Var{i}(1:40))
					disp(sprintf('some variable have been truncated'));
				else
					netcdf.putVar(groupID,varid,startpoint,endpoint,Var{i})
				end
			end
		end
	elseif isa(Var,'struct'),
		%Start by getting the structure fields and size
		locfields=fields(Var);
		[dims,DimSize,DimValue]=GetDims(ncid,Var,DimSize,DimValue);
		varid = netcdf.defVar(groupID,field,'NC_CHAR',dims);
		if length(locfields)==0,
			netcdf.putVar(groupID,varid,[0,0],[1,11],'emptystruct')
		else
			for i=1:length(locfields),
				for j=1:2,
					if j==1,
						CharVar=locfields{i};
						if length(CharVar)==0
							CharVar='emptystruct';
						end
						startpoint=[i-1,0,0];
					else
						if isa(Var.(locfields{i}),'char'),
							CharVar=Var.(locfields{i});
						else
							CharVar=num2str(Var.(locfields{i}));
						end
						if length(CharVar)==0
							CharVar='emptystruct';
						end
						startpoint=[i-1,1,0];
					end

					extent=[1,1,min(length(CharVar),40)];
					if length(CharVar)>40,
						netcdf.putVar(groupID,varid,startpoint,extent,CharVar(1:40))
						disp(sprintf('some variable have been truncated'));
					else
						netcdf.putVar(groupID,varid,startpoint,extent,CharVar)
					end
				end
			end
		end
	else
		disp(sprintf('no support for class %s of field %s',varclass,field));
  end
	return
end

function [dims,DimSize,DimValue]=GetDims(ncid,Var,DimSize,DimValue)
	dims=[];
	if isa(Var,'cell'),
		varsize=size(Var');
	elseif isa(Var,'struct')
		varsize=length(fields(Var));
	else
		varsize=size(Var);
	end
	dim=sum(varsize>1);
	if dim>0
		for i=1:dim
			indsize=find(varsize(i)==DimValue);
			if length(indsize)>0
				dims=[dims DimSize(indsize).index];
			else
				indsize=length(DimSize)+1;
				DimSize(indsize).index=netcdf.defDim(ncid,['Dimension' num2str(indsize)],varsize(i));
				[DimSize(indsize).name,DimSize(indsize).value]=netcdf.inqDim(ncid,DimSize(indsize).index);
				DimValue(indsize)=DimSize(indsize).value;
				dims=[dims DimSize(indsize).index];
			end
		end
	end
	%if we have a cell variable we need to add a stringlength dimension 
	if isa(Var,'struct'),
		if DimValue(3)~=2
			if DimValue(2)~=2
				dims=[dims DimSize(1).index];
			else
				dims=[dims DimSize(2).index];
			end
		else
			dims=[dims DimSize(3).index];
		end
	end
	if isa(Var,'cell') || isa(Var,'struct'),
		if DimValue(2)~=40
			dims=[dims DimSize(1).index];
		else
			dims=[dims DimSize(2).index];
		end
	end
end

File: trunk/src/m/contrib/defleurian/netCDF/export_netCDF.py
from netCDF4 import Dataset, stringtochar
import numpy as np
import time
import collections
from mesh2d import *
from mesh3dprisms import *
from results import *
from os import path, remove

def export_netCDF(md,filename):
	#Now going on Real treatment
	if path.exists(filename):
		print ('File {} allready exist'.format(filename))
		newname=raw_input('Give a new name or "delete" to replace: ')
		if newname=='delete':
			remove(filename)
		else:
			print ('New file name is {}'.format(newname))
			filename=newname
			
	NCData=Dataset(filename, 'w', format='NETCDF4')
	NCData.description = 'Results for run' + md.miscellaneous.name
	NCData.history = 'Created ' + time.ctime(time.time())

	#define netCDF dimensions
	try:
		StepNum=np.shape(dict.values(md.results.__dict__))[1]
	except IndexError:
		StepNum=1
	Dimension1=NCData.createDimension('DimNum1',StepNum)#time is first
	DimDict={len(Dimension1):'DimNum1'}
	dimindex=1

	dimlist=[2,md.mesh.numberofelements,md.mesh.numberofvertices,np.shape(md.mesh.elements)[1]]
	for i in range(0,4):
		if dimlist[i] not in DimDict.keys():
			dimindex+=1
			NewDim=NCData.createDimension('DimNum'+str(dimindex),dimlist[i])
			DimDict[len(NewDim)]='DimNum'+str(dimindex)

	typelist=[bool,str,unicode,int,float,complex,
						collections.OrderedDict,
						np.int64,np.ndarray,np.float64]
	groups=dict.keys(md.__dict__)
	#get all model classes and create respective groups
	for group in groups:
		NCgroup=NCData.createGroup(str(group))
		#In each group gather the fields of the class
		fields=dict.keys(md.__dict__[group].__dict__)

		#looping on fields
		for field in fields:
			#Special treatment for list fields
			if type(md.__dict__[group].__dict__[field])==list:
				StdList=False
				if len(md.__dict__[group].__dict__[field])==0:
					StdList=True
				else:
					StdList=type(md.__dict__[group].__dict__[field][0]) in typelist
				NCgroup.__setattr__('classtype', md.__dict__[group].__class__.__name__)
				if StdList: #this is a standard or empty list just proceed
					Var=md.__dict__[group].__dict__[field]
					DimDict=CreateVar(NCData,Var,field,NCgroup,DimDict)
				else: #this is a list of fields, specific treatment needed
					Listsize=len(md.__dict__[group].__dict__[field])
					Subgroup=NCgroup.createGroup(str(field))
					Subgroup.__setattr__('classtype',md.__dict__[group].__dict__[field].__class__.__name__)
					for listindex in range(0,Listsize):
						try:
							Listgroup=Subgroup.createGroup(str(md.__dict__[group].__dict__[field].__getitem__(listindex).__dict__['name']))
						except KeyError:
							for naming in ['step']:
								Listgroup=Subgroup.createGroup(str(md.__dict__[group].__dict__[field].__getitem__(listindex).__dict__[naming]))
						except AttributeError:
							Listgroup=Subgroup.createGroup(str(md.__dict__[group].__dict__[field].__class__.__name__)+str(listindex))
						Listgroup.__setattr__('classtype',md.__dict__[group].__dict__[field].__getitem__(listindex).__class__.__name__)
						try:
							subfields=dict.keys(md.__dict__[group].__dict__[field].__getitem__(listindex).__dict__)
						except AttributeError:
							subfields=dict.keys(md.__dict__[group].__dict__[field].__getitem__(listindex))
						for subfield in subfields:
							if subfield!='outlog':
								try:
									Var=md.__dict__[group].__dict__[field].__getitem__(listindex).__dict__[subfield]
								except AttributeError:
									Var=md.__dict__[group].__dict__[field].__getitem__(listindex)[subfield]
								DimDict=CreateVar(NCData,Var,subfield,Listgroup,DimDict,md.__dict__[group],field,listindex)

			#No subgroup, we directly treat the variable
			elif type(md.__dict__[group].__dict__[field]) in typelist or field=='bamg':
				NCgroup.__setattr__('classtype', md.__dict__[group].__class__.__name__)
				Var=md.__dict__[group].__dict__[field]
				DimDict=CreateVar(NCData,Var,field,NCgroup,DimDict)
			elif md.__dict__[group].__dict__[field] is None:
				print( 'field md.{}.{} is None'.format(group,field))
				#do nothing
			else:
				NCgroup.__setattr__('classtype', str(group))
				Subgroup=NCgroup.createGroup(str(field))
				Subgroup.__setattr__('classtype',md.__dict__[group].__class__.__name__)
				subfields=dict.keys(md.__dict__[group].__dict__[field].__dict__)
                                
				for subfield in subfields:
					if str(subfield)!='outlog':
						Var=md.__dict__[group].__dict__[field].__dict__[subfield]
						DimDict=CreateVar(NCData,Var,subfield,Subgroup,DimDict)
				
	NCData.close()

#============================================================================
#Define the variables
def CreateVar(NCData,var,field,Group,DimDict,*step_args):
	#grab type
	try:
		val_type=str(var.dtype)
	except AttributeError:
		val_type=type(var)
		#grab dimension
	try:
		val_shape=dict.keys(var)
	except TypeError:
		val_shape=np.shape(var)

	TypeDict = {float:'f8',
							'float64':'f8',
							np.float64:'f8',
							int:'i8',
							'int64':'i8',
							np.int64:'i8',
							str:str,
							dict:str}
		
	val_dim=np.shape(val_shape)[0]
	#Now define and fill up variable
	#treating scalar string or bool as atribute
	if val_type==str or val_type==unicode or val_type==bool:
		Group.__setattr__(str(field).swapcase(), str(var))
	#treating list as string table
	elif val_type==list:
		dimensions,DimDict=GetDim(NCData,var,val_shape,DimDict,val_dim)
		#try to get the type from the first element
		try:
			nctype=TypeDict[type(var[0])]
		except IndexError:
			nctype=str #most probably an empty list take str for that
		ncvar = Group.createVariable(str(field),nctype,dimensions,zlib=True)
		if val_shape==0:
			ncvar= []
		else:			
			for elt in range(0,val_shape[0]):
				ncvar[elt] = var[elt]
	#treating bool tables as string tables
	elif val_type=='bool':
		dimensions,DimDict=GetDim(NCData,var,val_shape,DimDict,val_dim)
		ncvar = Group.createVariable(str(field),str,dimensions,zlib=True)
		for elt in range(0,val_shape[0]):
			ncvar[elt] = str(var[elt])
	#treating dictionaries as tables of strings
	elif val_type==collections.OrderedDict or val_type==dict:
		dimensions,DimDict=GetDim(NCData,var,val_shape,DimDict,val_dim)
		ncvar = Group.createVariable(str(field),str,dimensions,zlib=True)
		for elt in range(0,val_dim):
			ncvar[elt,0]=dict.keys(var)[elt]
			ncvar[elt,1]=str(dict.values(var)[elt]) #converting to str to avoid potential problems
	#Now dealing with numeric variables
	else:
		dimensions,DimDict=GetDim(NCData,var,val_shape,DimDict,val_dim)
		ncvar = Group.createVariable(str(field),TypeDict[val_type],dimensions,zlib=True)
		try:
			nan_val=np.isnan(var)
			if nan_val.all():
				ncvar [:] = 'NaN'
			else:
				ncvar[:] = var
		except TypeError: #type does not accept nan, get vallue of the variable
			ncvar[:] = var
	return DimDict

#============================================================================
#retriev the dimension tuple from a dictionnary
def GetDim(NCData,var,shape,DimDict,i):
	output=[]
	#grab dimension
	for dim in range(0,i): #loop on the dimensions
		if type(shape[0])==int:
			try:
				output=output+[str(DimDict[shape[dim]])] #test if the dimension allready exist
			except KeyError: #if not create it
				if (shape[dim])>0:
					index=len(DimDict)+1
					NewDim=NCData.createDimension('DimNum'+str(index),(shape[dim]))
					DimDict[len(NewDim)]='DimNum'+str(index)
					output=output+[str(DimDict[shape[dim]])]
		elif type(shape[0])==str or type(shape[0])==unicode:#dealling with a dictionnary
			try:
				#dimension5 is 2 to treat with dict
				output=[str(DimDict[np.shape(shape)[0]])]+[DimDict[2]]
			except KeyError:
				index=len(DimDict)+1
				NewDim=NCData.createDimension('DimNum'+str(index),np.shape(shape)[0])
				DimDict[len(NewDim)]='DimNum'+str(index)
				output=[str(DimDict[np.shape(dict.keys(var))[0]])]+[DimDict[2]]
			break
	return tuple(output), DimDict
